### PR TITLE
Bbsr cluster with refactor merged in

### PR DESCRIPTION
**Pull request by NicoDeVeaux**

### CODE REVIEW

**Comment by NicoDeVeaux** [comment]
@tymor22 Python3.4 is giving unexpected test failures

File: inferelator_ng/bbsr_python.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+        betas = np.repeat(0.0,K)
+        if best > 0:
+
+            lst_combos_bool=combos[:, best]
+            lst_true_index = [i for i, j in enumerate(lst_combos_bool) if j]
+            x_tmp = x[:,lst_true_index]
+
+            try:
+                bhat = np.linalg.solve(np.dot(x_tmp.transpose(),x_tmp),np.dot(x_tmp.transpose(),y))
+                for m in xrange(len(lst_true_index)):
```
Comment (AaronWatters) [inline inferelator_ng/bbsr_python.py]: Python 3 has no `xrange` -- change this to `range`.

File: inferelator_ng/bbsr_python.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+                    ind_t=lst_true_index[m]
+                    betas[ind_t]=bhat[m]
+                not_done = False
+
+            except:
```
Comment (AaronWatters) [inline inferelator_ng/bbsr_python.py]: What exception is expected here?  This unquailified except masked a NameError.  Please be more specific.

**Comment by AaronWatters** [comment]
Here is how to add a git link to a requests file (example)

-e git://github.com/simonsfoundation/jp_svg_canvas@1.3#egg=jp_svg_canvas

**Comment by kostyat** [comment]
I tried running nosetests on my machine, and got the following warnings:

```
../home/tchoko09/inferelator_ng/inferelator_ng/bbsr_python.py:299: FutureWarning: in the future, boolean array-likes will be handled as a boolean array index
  bs_betas.ix[res['ind'],X.index.values[res['pp']]] = res['betas']
/home/tchoko09/inferelator_ng/inferelator_ng/bbsr_python.py:300: FutureWarning: in the future, boolean array-likes will be handled as a boolean array index
  bs_betas_resc.ix[res['ind'],X.index.values[res['pp']]] = res['betas_resc']
../home/tchoko09/anaconda2/lib/python2.7/site-packages/numpy/core/_methods.py:82: RuntimeWarning: Degrees of freedom <= 0 for slice
  warnings.warn("Degrees of freedom <= 0 for slice", RuntimeWarning)
..............S................................................../home/tchoko09/inferelator_ng/inferelator_ng/tfa.py:44: UserWarning: 1 transcription factors are removed because no expression or prior information exists.
  warnings.warn(message)
..................
----------------------------------------------------------------------
Ran 87 tests in 11.100s

```

I haven't been able to run the code on Bsubtilis or Yeast, because I can't figure out how to install the KVS client as a module for python :/ I tried installing it using the command

`sudo pip install git+https://github.com/simonsfoundation/kvsstcp.git`

but when running

`bsubtilis_bbsr_workflow_runner.py`

I still got the error `**ImportError: No module named kvsclient**`

**Comment by NicoDeVeaux** [comment]
To add KVS client:

git clone https://github.com/simonsfoundation/kvsstcp
export PYTHONPATH=$PYTHONPATH:$(pwd)/kvsstcp

**Comment by NicoDeVeaux** [comment]
Review of 10/13/17 meeting:
1) Review and merge pending pull request
2) Unit Tests
3) Refactor (write new code) — Objects, parellelization, precomputing cross-products

**Comment by gs512** [comment]
While trying to recreate the same conditions as Kostya described during the meeting, and following the previous instructions ( kvs = None and rank = 0 ) I got the following error : 

python2 yeast_bbsr_workflow_runner.py
Creating design and response matrix ...
Setting up TFA specific response matrix ...
Computing Transcription Factor Activity ...
Bootstrap 1 of 2
Calculating MI, Background MI, and CLR Matrix
Traceback (most recent call last):
  File "yeast_bbsr_workflow_runner.py", line 12, in <module>
    workflow.run()
  File "/Users/gs/Desktop/test_bug/inferelator_ng/inferelator_ng/bbsr_tfa_workflow.py", line 48, in run
    kvs.put('mi %d'%idx, (self.clr_matrix, self.mi_matrix))
AttributeError: 'NoneType' object has no attribute 'put'

**Comment by kostyat** [comment]
By the way, our conclusion after digging into my error is that we need to reimplement the step that removed or grouped the predictors that each highly correlated with each other. Otherwise, we will probably get computational errors.

On the other hand, while we got the error when xtx was calculated first, and then multiplied by bhat on both sides, if we changed the calculation such that first x times bhat is computed, and then the norm of that vector is taken (which is the same thing mathematically), there was no longer a computational error (i.e., the norm came out to be a normal, two-digit number (87 to be precise), and not -6.7e13).

On the other hand, we need someone to confirm that this error is not only happening on my machine. We tried to get Giuseppe to do that on his mac, but he got the kvs error that he described (after setting something to None and rank to 0, just like Nick did on his mac).

File: inferelator_ng/bbsr_python.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        pp.ix[ind, clr_order[0:min(K, nS, len(clr_order))]] = True
+
+    preds = np.intersect1d(genes, tfs)
+    subset = pp.ix[preds,preds].values
+    np.fill_diagonal(subset,False)
+    pp=pp.set_value(preds, preds, subset)
+
+    out_list=[]
+
+    global gx, gy, gpp, gwm, gns
```
Comment (AaronWatters) [inline inferelator_ng/bbsr_python.py]: why globals?

File: inferelator_ng/bbsr_python.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    gx, gy, gpp, gwm, gns = X, Y, pp, weights_mat, nS
+    # Here we illustrate splitting a simple loop, but the same approach
+    # would work with any iterative control structure, as long as it is
+    # deterministic.
+    s = []
+    limit = G
+    for j in range(limit):
+        if not ownCheck or ownCheck.next():
+            # busy work
+            s.append(BBSRforOneGeneWrapper(j))
```
Comment (AaronWatters) [inline inferelator_ng/bbsr_python.py]: This removes the need for globals:
`s.append(BBSRforOneGene(j, X, Y, pp, weights_mat, nS))`

**Pull request merged**